### ISSUE TITLE
Variable minibatch size to optimally use GPU memory

### ISSUE THREAD

**Issue by Waino** [robertostling/hnmt#1]
Vary the number of sentences per minibatch, so that the number of tokens (including padding) is close to optimal. The current fixed-size minibatches under-utilize the available memory when the minibatch is filled with short sentences.

**Comment by robertostling** [comment]
Currently working on a solution for this (integrated with code for randomly iterating through larger-than-RAM data sets).

**Comment by Waino** [comment]
Did you notice the branch variable_minibatch ?
It contains a working solution for this issue, which still has some rough edges needing to be polished.
The new command line parameter ``--batch-budget`` controls the memory usage.

Profiling showed that encoding the sentences (from strings to numpy arrays) immediately before they are needed caused a heavy CPU load. I solved this by encoding the sentences to numpy arrays without padding at startup. 

The unfinished parts are:
* ``--batch-size`` is still needed for translation.
* The weights for the budget calculation are not tuned.
* bnas.TextEncoder needed to be modified. I copied the relevant parts into text.py in hnmt, but this is an ugly solution. Either the changes should be included in bnas, or hnmt should be made into a "proper" package.
